test/integration/conftest.py

```diff
@@ -0,0 +1,146 @@
+from __future__ import annotations
+
+import functools
+import logging
+import subprocess
+from typing import Callable, Generator
+
+import jubilant
+import minio
+import pytest
+
+
+@pytest.fixture
+def juju() -> Generator[jubilant.Juju, None, None]:
+    """Make a Juju model with the tracing part of COS ready."""
+    with jubilant.temp_model() as j:
+        deploy_tempo(j)
+        deploy_tempo_worker(j)
+        j.deploy('minio', config={'access-key': 'accesskey', 'secret-key': 'mysoverysecretkey'})
+        j.deploy('s3-integrator')
+
+        j.integrate('tempo:s3', 's3-integrator')
+        j.integrate('tempo:tempo-cluster', 'tempo-worker')
+
+        j.wait(lambda s: minio_active(s) and s3_integrator_blocked(s))
+
+        address = j.status().apps['minio'].address
+        mc_client = minio.Minio(
+            f'{address}:9000',
+            access_key='accesskey',
+            secret_key='mysoverysecretkey',
+            secure=False,
+        )
+
+        found = mc_client.bucket_exists('tempo')
+        if not found:
+            mc_client.make_bucket('tempo')
+
+        j.config('s3-integrator', dict(endpoint=f'http://{address}:9000', bucket='tempo'))
+        j.run(
+            's3-integrator/0',
+            'sync-s3-credentials',
+            {'access-key': 'accesskey', 'secret-key': 'mysoverysecretkey'},
+        )
+
+        # Tempo goes through a cycle of:
+        # - update own stateful set
+        # - kill own pod
+        # - new pod is scheduled by k8s
+        # - check own stateful set
+        #
+        # This process may take a while. I'm unsure about the default timeout.
+
+        j.wait(jubilant.all_active)
+
+        yield j
+
+        print(j.debug_log())
+
+
+@pytest.fixture(scope='session')
+def build_charm(
+    pytestconfig: pytest.Config,
+) -> Generator[Callable[[], str], None, None]:
+    """Build the test charm and provide the artefact path.
+
+    Starts building the tracing-tester charm early.
+    Call the fixture value to get the built charm file path.
+    """
+    charm_dir = pytestconfig.rootpath / 'test/charms/test_integration'
+    proc = subprocess.Popen(
+        ['charmcraft', 'pack'],  # noqa: S607
+        cwd=str(charm_dir),
+        universal_newlines=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+
+    def tracing_test_charm():
+        proc.communicate()
+        assert proc.returncode == 0
+        charm = charm_dir / 'tracing-tester_amd64.charm'
+        assert charm.exists()
+        return str(charm)
+
+    yield tracing_test_charm
+    if proc.returncode is None:
+        proc.kill()
+    out, err = proc.communicate()
+    logging.info('`charmcraft pack` stdout follows:\n%s', out)
+    logging.info('`charmcraft pack` stderr follows:\n%s', err)
+
+
+def app_is(s: jubilant.Status, app: str, status: str):
+    return next(iter(s.apps[app].units.values())).workload_status.current == status
```

Comment:
This `app_is` helper seems misnamed with what it actually does -- check *unit* status.

- Do we actually want to be testing unit status here? Would `jubilant.all_active` et al work? (They test both app and unit status)
- It seems clearer to me to put these as lambdas inline in the test. That way a person maintaining the test doesn't have to go two layers deeper to figure out what it actually does.